gateway/hydrator.go

```diff
@@ -164,13 +164,21 @@ func (eph *EndpointHydrator) performChecks(serviceID protocol.ServiceID, service
 				endpointLogger.Info().Msg("running checks against the endpoint")
 
 				// Retrieve all the required QoS checks for the endpoint.
-				requiredQoSChecks := serviceQoS.GetRequiredQualityChecks(endpoint.Addr())
+				requiredQoSChecks := serviceQoS.GetRequiredQualityChecks()
 				if len(requiredQoSChecks) == 0 {
 					endpointLogger.Warn().Msg("service QoS returned 0 required checks")
 					continue
 				}
 
 				for _, serviceRequestCtx := range requiredQoSChecks {
+					// Create a new protocol request context with a pre-selected endpoint for each request.
+					// This is to avoid race conditions or concurrent access problems when running concurrent QoS checks.
+					hydratorRequestCtx, err := eph.Protocol.BuildHydratorRequestContextForEndpoint(serviceID, endpoint.Addr())
```

Comment:
This should be simply named `BuildRequestContextForEndpoint`, protocol should not know which entity is going to use the returned context.